File: android/feature/azkar/src/main/kotlin/com/fatwabot/feature/azkar/AzkarModels.kt
package com.fatwabot.feature.azkar

import java.io.File
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/** Persisted mid-session position — mirror of iOS AzkarSessionState
 * (docs/features/azkar.md: "Resume mid-session after app restart same day"). */
@Serializable
data class AzkarSessionState(
    val categoryId: String,
    val currentItemIndex: Int,
    val currentItemCount: Int,
    val lastTouchedAtEpochSeconds: Long,
)

/** Local completion history — feeds "completed today" and, later, M3 streak events. */
@Serializable
data class AzkarCompletionRecord(
    val categoryId: String,
    val completedAtEpochSeconds: Long,
)

interface AzkarStoring {
    fun loadSession(): AzkarSessionState?
    fun saveSession(session: AzkarSessionState?)
    fun loadCompletions(): List<AzkarCompletionRecord>
    fun recordCompletion(record: AzkarCompletionRecord)
}

class FileAzkarStore(private val directory: File) : AzkarStoring {
    private val json = Json { ignoreUnknownKeys = true }
    private val sessionFile = File(directory, "azkar-session.json")
    private val completionsFile = File(directory, "azkar-completions.json")
    private val completionsSerializer = ListSerializer(AzkarCompletionRecord.serializer())

    override fun loadSession(): AzkarSessionState? = runCatching {
        if (!sessionFile.exists()) {
            null
        } else {
            json.decodeFromString(AzkarSessionState.serializer(), sessionFile.readText())
        }
    }.getOrNull()

    override fun saveSession(session: AzkarSessionState?) {
        if (session == null) {
            sessionFile.delete()
            return
        }
        runCatching {
            directory.mkdirs()
            sessionFile.writeText(json.encodeToString(AzkarSessionState.serializer(), session))
        }
    }

    override fun loadCompletions(): List<AzkarCompletionRecord> = runCatching {
        if (!completionsFile.exists()) {
            emptyList()
        } else {
            json.decodeFromString(completionsSerializer, completionsFile.readText())
        }
    }.getOrDefault(emptyList())

    override fun recordCompletion(record: AzkarCompletionRecord) {
        val all = loadCompletions() + record
        runCatching {
            directory.mkdirs()
            completionsFile.writeText(json.encodeToString(completionsSerializer, all))
        }
    }
}

File: ios/Packages/AppFeatures/Sources/AzkarFeature/AzkarModels.swift
import Foundation

/// Persisted mid-session position (docs/features/azkar.md: "Resume mid-session
/// after app restart same day restores exact position").
public struct AzkarSessionState: Codable, Equatable, Sendable {
    public let categoryId: String
    public var currentItemIndex: Int
    public var currentItemCount: Int
    public var lastTouchedAt: Date

    public init(categoryId: String, currentItemIndex: Int, currentItemCount: Int, lastTouchedAt: Date) {
        self.categoryId = categoryId
        self.currentItemIndex = currentItemIndex
        self.currentItemCount = currentItemCount
        self.lastTouchedAt = lastTouchedAt
    }
}

/// Local completion history — feeds the "completed today" badge and, later,
/// the M3 streak engine's activity events.
public struct AzkarCompletionRecord: Codable, Equatable, Sendable {
    public let categoryId: String
    public let completedAt: Date

    public init(categoryId: String, completedAt: Date) {
        self.categoryId = categoryId
        self.completedAt = completedAt
    }
}

public protocol AzkarStoring: Sendable {
    func loadSession() -> AzkarSessionState?
    func saveSession(_ session: AzkarSessionState?)
    func loadCompletions() -> [AzkarCompletionRecord]
    func recordCompletion(_ record: AzkarCompletionRecord)
}

public struct FileAzkarStore: AzkarStoring {
    private let sessionURL: URL
    private let completionsURL: URL

    public init(directory: URL) {
        sessionURL = directory.appendingPathComponent("azkar-session.json")
        completionsURL = directory.appendingPathComponent("azkar-completions.json")
    }

    public func loadSession() -> AzkarSessionState? {
        guard let data = try? Data(contentsOf: sessionURL) else { return nil }
        return try? JSONDecoder().decode(AzkarSessionState.self, from: data)
    }

    public func saveSession(_ session: AzkarSessionState?) {
        guard let session else {
            try? FileManager.default.removeItem(at: sessionURL)
            return
        }
        guard let data = try? JSONEncoder().encode(session) else { return }
        try? FileManager.default.createDirectory(
            at: sessionURL.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        try? data.write(to: sessionURL, options: .atomic)
    }

    public func loadCompletions() -> [AzkarCompletionRecord] {
        guard let data = try? Data(contentsOf: completionsURL) else { return [] }
        return (try? JSONDecoder().decode([AzkarCompletionRecord].self, from: data)) ?? []
    }

    public func recordCompletion(_ record: AzkarCompletionRecord) {
        var all = loadCompletions()
        all.append(record)
        guard let data = try? JSONEncoder().encode(all) else { return }
        try? FileManager.default.createDirectory(
            at: completionsURL.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        try? data.write(to: completionsURL, options: .atomic)
    }
}
